control-plane/agents/src/bin/core/tests/volume/snapshot.rs

```diff
@@ -417,3 +417,151 @@ async fn snapshot_timeout() {
+        .publish(
+            &PublishVolume {
+                uuid: volume.uuid().clone(),
+                ..Default::default()
+            },
+            None,
+        )
+        .await
+        .unwrap();
+
+    // Create a nexus snapshot.
+    vol_cli
+        .create_snapshot(
+            &CreateVolumeSnapshot::new(volume.uuid(), parent_nexus_snapid.clone()),
+            None,
+        )
+        .await
+        .unwrap();
+
+    // Stop the core agent so that we can create the leaked resources.
+    cluster.composer().stop("core").await.unwrap();
+
+    // Extract the nexus and replica to be used to create leaked resources.
+    let volume_state = volume.state();
+    let replica = volume_state.replica_topology.iter().next().unwrap();
+    let nexus = volume_state.target.unwrap();
+
+    // Create the rpc handle.
+    let mut rpc_handle = cluster
+        .grpc_handle(&cluster.node(0))
+        .await
+        .expect("Should get handle");
+    // At this point no leaked resources create so, snapshot count is 2.
+    let snaps = rpc_handle
+        .list_replica_snaps(None, None)
+        .await
+        .expect("Should get replica snaps");
+    assert_eq!(snaps.snapshots.len(), 2);
+
+    // Create the first leaked replica snapshot, with the existing replica snapshot's id.
+    rpc_handle
+        .create_replica_snap(
+            &format!("{parent_repl_snapid}/322"),
+            volume.spec().uuid.as_str(),
+            "322",
+            replica.0.as_str(),
+            SnapshotId::new().as_str(),
+        )
+        .await
+        .unwrap();
+    let snaps = rpc_handle
+        .list_replica_snaps(None, None)
+        .await
+        .expect("Should get replica snaps");
+    assert_eq!(snaps.snapshots.len(), 3);
+
+    // Create the first leaked replica snapshot, with the existing nexus snapshot's id.
+    rpc_handle
+        .create_nexus_snap(
+            nexus.uuid.as_str(),
+            &format!("{parent_nexus_snapid}/766"),
+            volume.spec().uuid.as_str(),
+            "766",
+            replica.0.as_str(),
+            SnapshotId::new().as_str(),
+        )
+        .await
+        .unwrap();
+    let snaps = rpc_handle
+        .list_replica_snaps(None, None)
+        .await
+        .expect("Should get replica snaps");
+    assert_eq!(snaps.snapshots.len(), 4);
+
+    // Create the third leaked replica snapshot, with the snapshot id that no longer exists.
+    rpc_handle
+        .create_nexus_snap(
+            nexus.uuid.as_str(),
+            &format!("{}/54", SnapshotId::new()),
+            volume.spec().uuid.as_str(),
+            "54",
+            replica.0.as_str(),
+            SnapshotId::new().as_str(),
+        )
+        .await
+        .unwrap();
+    let snaps = rpc_handle
+        .list_replica_snaps(None, None)
+        .await
+        .expect("Should get replica snaps");
+    assert_eq!(snaps.snapshots.len(), 5);
+
+    cluster.composer().start("core").await.unwrap();
```

Comment:
Isn't there a restart core in the cluster? Which does the wait as well

Reply:
restart_core seems to restart only and not wait.

Reply:
Ah, it removes the lock, we should use restart followed by the liveness then.
Maybe we can add a helper method that does both?

Reply:
Done. Added the helper.